src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -210,26 +216,70 @@ KafkaSink::KafkaSink(
     });
 }
 
-void KafkaSink::addMessageToBatch(char * pos, size_t len)
+/// When this function is called, there could be two scenarios:
+/// 1) len == total_len. That means the whole Chunk (or the whole row, when one_message_per_row is true) can fit into the buffer.
+///    In this case, we just need to create a rd_kafka_message_t, and push it to current_batch.
+///
+/// 2) len < total_len. That means the Chunk (or a row) is too big to fit into the buffer. In this case, the data between
+///    [pos + len, pos + total_len] will be copied into pending_data so that the bufer can accept more data until it gets
+///    complete data. Then it can create a rd_kafka_message_t.
+///
+/// In this way, we limit the size of one single Kafka message, so that it can avoid hitting the max.message.size.
+/// However, it's still possible that, one single row is still too big and it exceeds that limit. There is nothing we can do about it for now.
+void KafkaSink::addMessageToBatch(char * pos, size_t len, size_t total_len)
```

Comment:
if `max_ingest_block_size` means row count in the same kafka message, then i think things may be easier here

Reply:
See above comments.